dali/base/dasds.cpp

```diff
@@ -1015,72 +1044,161 @@ void writeDelta(StringBuffer &xml, IFile &iFile, const char *msg="", unsigned re
     {
-        delete text;
-        if (data) free(data);
+        type = f_delta;
+        name = path;
+    }
+    CTransactionItem(char *_name, size32_t _dataLength, void *_data) : name(_name), dataLength(_dataLength), data(_data)
+    {
+        type = data ? f_addext : f_delext;
+    }
+    ~CTransactionItem()
+    {
+        free(name);
+        switch (type)
+        {
+            case f_delta:
+                ::Release(deltaTree);
+                break;
+            case f_addext:
+                free(data);
+                break;
+            case f_delext:
+            default:
+                break;
+        }
     }
-    StringBuffer *text;
-    unsigned edition;
-    unsigned dataLength;
-    void *data;
-    byte flags;
 };
-unsigned BackupQueueItem::typeMask = 0x0f;
-class CBackupHandler : public CInterface, implements IThreaded
-{
-    typedef QueueOf<BackupQueueItem, false> BackupQueue;
-    CThreaded threaded;
-    BackupQueue itemQueue, freeQueue;
-    Semaphore pending, softQueueLimitSem;
-    bool aborted, waiting, addWaiting, async;
-    unsigned currentEdition, throttleCounter;
-    CriticalSection queueCrit, freeQueueCrit;
-    StringAttr backupPath;
-    unsigned freeQueueLimit;  // how many BackupQueueItems to cache for reuse
-    unsigned largeWarningThreshold;    // point at which to start warning about large queue
-    unsigned softQueueLimit; // threshold over which primary transactions will be delay by small delay, to allow backup catchup.
-    unsigned softQueueLimitDelay; // delay for above
-    CTimeMon warningTime;
-    unsigned recentTimeThrottled;
-    unsigned lastNumWarnItems;
-    IPropertyTree &config;
 
-    const unsigned defaultFreeQueueLimit = 50;
-    const unsigned defaultLargeWarningThreshold = 50;
-    const unsigned defaultSoftQueueLimit = 200;
-    const unsigned defaultSoftQueueLimitDelay = 200;
+static constexpr unsigned defaultSaveThresholdSecs = 0; // disabled
+static constexpr unsigned defaultDeltaSaveTransactionThreshold = 0; // disabled
+static constexpr unsigned defaultDeltaMemMaxMB = 10;
+static constexpr unsigned defaultDeltaTransactionQueueLimit = 10000;
+class CDeltaWriter : implements IThreaded
+{
+    IStoreHelper *iStoreHelper = nullptr;
+    StringBuffer dataPath;
+    StringBuffer backupPath;
+    unsigned transactionQueueLimit = defaultDeltaTransactionQueueLimit; // absolute limit, will block if this far behind
+    memsize_t transactionMaxMem = defaultDeltaMemMaxMB * 0x100000; // 10MB
+    unsigned totalQueueLimitHits = 0;
+    unsigned addQueueWaiting = 0;
+    unsigned saveThresholdSecs = 0;
+    cycle_t lastSaveTime = 0;
+    cycle_t thresholdDuration = 0;
+
+    std::queue<Owned<CTransactionItem>> pending;
+    memsize_t pendingSz = 0;
+    CriticalSection pendingCrit;
+    CCycleTimer timer;
+    StringBuffer deltaXml;
+    CThreaded threaded;
+    Semaphore sem;
+    cycle_t timeThrottled = 0;
+    unsigned throttleCounter = 0;
+    bool waiting = true;
```

Comment:
This needs to be atomic

Reply:
does it? It's always altered or read whilst pendingCrit is held.